### DISCUSSION TITLE
Upgrading RabbitMQ and Erlang - Ordering and errors after upgrading

### DISCUSSION THREAD

**Discussion by DMahoney0** [rabbitmq/rabbitmq-server#2764] [Questions] [answered]
I have a machine that I've used WIX to install RabbitMQ 3.7.3 and Erlang 20.0 in the past. Now I want to upgrade to RabbitMQ 3.7.28 or later to get the patch fixes.

I've tried upgrading RabbitMQ to 3.7.28 first and then installing Erlang 21.3 according to the [Compatibility Matrix](https://www.rabbitmq.com/which-erlang.html) and [the upgrade instructions](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.7.28)

Order of installation:
- Erlang 20.0 (in the past / previous release)
- RabbitMQ 3.7.3 (in the past / previous release)
- Upgrade to RabbitMQ 3.7.28 (today)
- Upgrade to Erlang 21.3 (today)

I've also tried upgrading to the latest RabbitMQ, 3.8.11, with Erlang 23.2.

Order of installation:
- Erlang 20.0 (in the past / previous release)
- RabbitMQ 3.7.3 (in the past / previous release)
- Upgrade to RabbitMQ 3.8.11 (today)
- Upgrade to Erlang 23.2 (today)

In both cases, I am getting a error when running the "rabbitmqctl.bat status" command after completely the upgrades:
`11:48:41.291 [error] Loading of c:/Program Files (x86)/RabbitMQ Server/rabbitmq_server-3.7.28/escript/rabbitmqctl/Elixir.RabbitMQCtl.beam failed: :badfile

11:48:41.295 [error] beam/beam_load.c(1862): Error loading module 'Elixir.RabbitMQCtl':
  This BEAM file was compiled for a later version of the run-time system than 20.
  To fix this, please recompile this module with an 20 compiler.
  (Use of opcode 162; this emulator supports only up to 159.)

** (UndefinedFunctionError) function RabbitMQCtl.main/1 is undefined (module RabbitMQCtl is not available)
    (rabbitmqctl) RabbitMQCtl.main(["status"])
    (elixir) lib/kernel/cli.ex:105: anonymous fn/3 in Kernel.CLI.exec_fun/2`

Any help on how I can debug this or if I am doing something wrong with the ordering would be helpful!

**Comment by lukebakken** [comment, marked as answer]
You must uninstall the previous version of Erlang before installing a new Erlang and RabbitMQ version. This is due to the fact that the RabbitMQ `.exe` installer searches for the Erlang installation in a way that sometimes selects the old version as the one to use.

Please see the following documentation:

https://www.rabbitmq.com/install-windows.html#upgrading-erlang (Note: we need to add a caveat about uninstalling Erlang - https://github.com/rabbitmq/rabbitmq-website/issues/1136)

https://www.rabbitmq.com/windows-quirks.html#multiple-erlang-versions

**Reply by DMahoney0** [reply]
@lukebakken do you know the command to uninstall the broker? 

Taken from: "If you have an existing installation and are planning to upgrade the Erlang VM from a 32-bit to a 64-bit version then you must uninstall the broker before upgrading the VM. The installer will not be able to stop or remove a service that was installed with an Erlang VM of a different architecture."

I've tried "rabbitmq-service.bat stop" and then "rabbitmq-service.bat remove" but not getting far with that. Is there another uninstall command that you know of? Perhaps something with rabbitmqctl?

**Reply by lukebakken** [reply]
The `.exe` installer provides uninstall function. You should also be able to find an `uninstall.exe` file in the installation directory.
